src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs

```diff
@@ -4698,6 +4698,298 @@ string HelloName(string name)
+
+            ValueTask ValueTaskMethod()
+            {
+                return ValueTask.CompletedTask;
+            }
+
+            Task TaskMethod()
+            {
+                return Task.CompletedTask;
+            }
+
+            async ValueTask ValueTaskWithYieldMethod()
+            {
+                await Task.Yield();
+            }
+
+            async Task TaskWithYieldMethod()
+            {
+                await Task.Yield();
+            }
+
+            return new object[][]
+            {
+                new object[] { (Action)VoidMethod },
+                new object[] { (Func<ValueTask>)ValueTaskMethod },
+                new object[] { (Func<Task>)TaskMethod },
+                new object[] { (Func<ValueTask>)ValueTaskWithYieldMethod },
+                new object[] { (Func<Task>)TaskWithYieldMethod}
+            };
+        }
+    }
+
+    [Theory]
+    [MemberData(nameof(VoidReturningMethods))]
+    public async Task CanInvokeFilter_OnVoidReturningHandler(Delegate @delegate)
+    {
+        // Arrange
+        var responseBodyStream = new MemoryStream();
+        var httpContext = CreateHttpContext();
+        httpContext.Response.Body = responseBodyStream;
+
+        // Act
+        var factoryResult = RequestDelegateFactory.Create(@delegate, new RequestDelegateFactoryOptions()
+        {
+            RouteHandlerFilterFactories = new List<Func<RouteHandlerContext, RouteHandlerFilterDelegate, RouteHandlerFilterDelegate>>()
+            {
+                (routeHandlerContext, next) => async (context) =>
+                {
+                    return await next(context);
+                }
+            }
+        });
+        var requestDelegate = factoryResult.RequestDelegate;
+        await requestDelegate(httpContext);
+
+        Assert.Equal(200, httpContext.Response.StatusCode);
+        var decodedResponseBody = Encoding.UTF8.GetString(responseBodyStream.ToArray());
+        Assert.Equal(String.Empty, decodedResponseBody);
+    }
+
+    public static object[][] AwaitedTaskReturningMethods
+    {
+        get
+        {
+            async Task HandlerWithTaskAwait(HttpContext c)
+            {
+                await Task.Delay(1000);
+                c.Response.StatusCode = 400;
+            };
+
+            async ValueTask HandlerWithValueTaskAwait(HttpContext c)
+            {
+                await Task.Delay(1000);
+                c.Response.StatusCode = 400;
+            };
+
+            return new object[][]
+            {
+                new object[] { (Func<HttpContext, Task>)HandlerWithTaskAwait },
+                new object[] { (Func<HttpContext, ValueTask>)HandlerWithValueTaskAwait },
+            };
+        }
+    }
+
+    [Fact]
+    public async Task CanInvokeFilter_OnTaskModifyingHttpContext()
+    {
+        // Arrange
+        var tcs = new TaskCompletionSource();
+        async Task HandlerWithTaskAwait(HttpContext c)
+        {
+            await tcs.Task;
```

Comment:
I don't think this is guaranteed to test the scenario, the TCS is completed before entering the handler. Thinking about it some more it's kind of hard to test this reliably... best I can think of is to set the TCS after `requestDelegate(httpContext);` and maybe adding a `Task.Yield()` after `await tcs.Task;`. Basically trying to force the delegate to be async and verifying that the `requestDelegate` doesn't complete synchronously.
So:
```csharp
var task = requestDelegate(httpContext);
tcs.SetResult();
await task;
```